src/utils/auth/auth.provider.tsx

```diff
@@ -0,0 +1,481 @@
+                targetAuthRequest.transaction as SplitTransaction,
+                chunksByCollectionID[collectionID]
+              )
+            );
+
+            const authRequests = prevAuthRequests.map((authRequest) => {
+              if (authRequest.authID !== targetAuthRequest.authID)
+                return authRequest;
+
+              return {
+                ...authRequest,
+                transaction
+              } as SignAuthRequest;
+            });
+
+            return {
+              authRequests,
+              currentAuthRequestIndex,
+              lastCompletedAuthRequest
+            };
+          }
+
+          // Update SignKeystoneAuthRequest with `data`:
+
+          const bytes = bytesFromChunks(chunksByCollectionID[collectionID]);
+          const data = Buffer.from(bytes);
+
+          const authRequests = prevAuthRequests.map((authRequest) => {
+            if (authRequest.authID !== targetAuthRequest.authID)
+              return authRequest;
+
+            return {
+              ...authRequest,
+              data
+            } as SignKeystoneAuthRequest;
+          });
+
+          return {
+            authRequests,
+            currentAuthRequestIndex,
+            lastCompletedAuthRequest
+          };
+        });
+      } else {
+        if (!chunksByCollectionID[collectionID]) {
+          chunksByCollectionID[collectionID] = [data];
+        } else {
+          chunksByCollectionID[collectionID].push(data);
+        }
+      }
+    });
+  }, []);
+
+  useEffect(() => {
+    let timeoutID = 0;
+
+    const isDone =
+      authRequests.length > 0 &&
+      authRequests.every((authRequest) => authRequest.status !== "pending");
+
+    if (initialScreenType === "default" && authRequests.length === 0) {
+      // Close the popup if an AuthRequest doesn't arrive in less than `AUTH_POPUP_REQUEST_WAIT_MS` (1s), unless the
+      // wallet is locked (no timeout in that case):
+      timeoutID = setTimeout(() => {
+        window.top.close();
+      }, AUTH_POPUP_REQUEST_WAIT_MS);
+    } else if (initialScreenType !== "default") {
+      // If the user doesn't unlock the wallet in 15 minutes, or somehow the popup gets stuck into any other state for
+      // more than that, we close it:
+      timeoutID = setTimeout(() => {
+        window.top.close();
+      }, AUTH_POPUP_UNLOCK_REQUEST_TTL_MS);
+    } else if (isDone) {
+      // Close the window if the last request has been handled:
+
+      // TODO: Add setting to decide whether this closes automatically or stays open in a "done" state.
+
+      if (AUTH_POPUP_CLOSING_DELAY_MS > 0) {
+        timeoutID = setTimeout(() => {
+          window.top.close();
+        }, AUTH_POPUP_CLOSING_DELAY_MS);
+      } else {
+        window.top.close();
+      }
+    }
+
+    function handleBeforeUnload() {
+      // Send cancel event for all pending requests if the popup is closed by the user:
+
+      authRequests.forEach((authRequest) => {
```

Comment:
We should add an issue to add a prompt for the user to confirm closing if they have multiple pending sign requests

Reply:
https://github.com/arconnectio/ArConnect/issues/562